File: kotlin-node/karakum/src/jsMain/resources/plugins/EventEmitterPlugin.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

function extractModifiers(member) {
    if (
        ts.isPropertyDeclaration(member)
        || ts.isMethodDeclaration(member)
        || ts.isConstructorDeclaration(member)
        || ts.isGetAccessorDeclaration(member)
        || ts.isSetAccessorDeclaration(member)
        || ts.isIndexSignatureDeclaration(member)
    ) {
        return Array.from(member.modifiers ?? [])
    }

    return []
}

function isEventEmitterInterfaceNode(node) {
    return (
        ts.isInterfaceDeclaration(node)
        && node.name.text === "EventEmitter"

        && node.parent
        && ts.isModuleBlock(node.parent)

        && node.parent.parent
        && ts.isModuleDeclaration(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && node.parent.parent.name.text === "NodeJS"
    )
}

function isEventEmitterClassNode(node) {
    return (
        ts.isClassDeclaration(node)
        && node.name
        && node.name.text === "EventEmitter"
    )
}

export default {
    setup(context) {
        this.eventEmitterInterfaceNode = null
    },

    traverse(node) {
        if (isEventEmitterInterfaceNode(node)) {
            this.eventEmitterInterfaceNode = node
        }
    },

    render(node, context, next) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
        if (!sourceFileName.endsWith("events.d.ts")) return null

        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)

        if (
            ts.isTypeReferenceNode(node)
            && ts.isIdentifier(node.typeName)
            && node.typeName.text === "Key"
        ) {
            return "EventType"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeReferenceNode(node.types[1])
            && ts.isIdentifier(node.types[1].typeName)
            && node.types[1].typeName.text === "Key2"
        ) {
            return "EventType"
        }

        if (
            ts.isTypeReferenceNode(node)
            && ts.isIdentifier(node.typeName)
            && (
                node.typeName.text === "Listener1"
                || node.typeName.text === "Listener2"
            )
        ) {
            return "EventListener"
        }

        if (
            ts.isIdentifier(node)
            && node.text === "eventName"

            && node.parent
            && ts.isParameter(node.parent)
            && node.parent.name === node

            && node.parent.parent
            && (
                ts.isMethodDeclaration(node.parent.parent)
                || ts.isMethodSignature(node.parent.parent)
            )
        ) {
            return "type"
        }

        if (
            ts.isMethodDeclaration(node)

            && node.parent
            && isEventEmitterClassNode(node.parent)
        ) {
            if (ts.isComputedPropertyName(node.name)) {
                return `/* ${typeScriptService?.printNode(node)} */`
            }

            const name = next(node.name)

            const returnType = node.type && next(node.type)

            return karakum.convertParameterDeclarations(node, context, next, {
                strategy: "function",
                template: (parameters, signature) => {
                    if (signature.some(it => it.type.kind === ts.SyntaxKind.SymbolKeyword)) {
                        return ""
                    }

                    if (
                        signature.length === 3
                        && ts.isTypeReferenceNode(signature[0].type)
                        && ts.isQualifiedName(signature[0].type.typeName)
                        && signature[0].type.typeName.right.text === "EventEmitter"
                    ) {
                        const enhancedReturnType = name === "once"
                            ? "Promise<P>"
                            : "AsyncIterator<P>"

                        return `fun <P : JsTuple> ${name}(emitter: EventEmitter, type: EventType, options: StaticEventEmitterOptions = definedExternally)${karakum.ifPresent(enhancedReturnType, it => `: ${it}`)}`
                    }

                    if (
                        signature.length === 3
                        && ts.isTypeReferenceNode(signature[0].type)
                        && ts.isIdentifier(signature[0].type.typeName)
                        && signature[0].type.typeName.text === "EventTarget"
                    ) {
                        const enhancedReturnType = name === "once"
                            ? "Promise<JsTuple1<E>>"
                            : "AsyncIterator<JsTuple1<E>>"

                        return `fun <E : Event> ${name}(emitter: EventTarget, type: web.events.EventType<E>, options: StaticEventEmitterOptions = definedExternally)${karakum.ifPresent(enhancedReturnType, it => `: ${it}`)}`
                    }

                    if (
                        signature.length === 2
                        && ts.isTypeReferenceNode(signature[0].type)
                        && ts.isQualifiedName(signature[0].type.typeName)
                        && signature[0].type.typeName.right.text === "EventEmitter"
                    ) {
                        return `fun ${name}(emitter: EventEmitter, type: EventType)${karakum.ifPresent(returnType, it => `: ${it}`)}`
                    }

                    if (
                        signature.length === 2
                        && ts.isTypeReferenceNode(signature[0].type)
                        && signature[0].type.typeName.text === "EventTarget"
                    ) {
                        return `fun ${name}(emitter: EventTarget, type: web.events.EventType<*>)${karakum.ifPresent(returnType, it => `: ${it}`)}`
                    }

                    // remove generics
                    return `fun ${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
                }
            })
        }

        if (
            ts.isMethodSignature(node)

            && node.parent
            && isEventEmitterInterfaceNode(node.parent)
        ) {
            if (ts.isComputedPropertyName(node.name)) {
                return `/* ${typeScriptService?.printNode(node)} */`
            }

            const name = next(node.name)

            const returnType = node.type && next(node.type)

            return karakum.convertParameterDeclarations(node, context, next, {
                strategy: "function",
                template: parameters => {
                    // remove generics
                    return `fun ${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
                }
            })
        }

        if (
            ts.isClassDeclaration(node)
            && node.name
            && node.name.text === "EventEmitter"
        ) {
            const heritageClauses = node.heritageClauses
                ?.map(heritageClause => next(heritageClause))
                ?.join(", ")

            const members = [
                ...node.members
                    .filter(member => extractModifiers(member).every(it => it.kind !== ts.SyntaxKind.StaticKeyword)),

                ...this.eventEmitterInterfaceNode.members,
            ]
                .map(member => next(member))
                .join("\n")

            const staticMembers = node.members
                .filter(member => extractModifiers(member).some(it => it.kind === ts.SyntaxKind.StaticKeyword))
                .map(member => next(member))
                .join("\n")

            const companionObject = `
companion object {
${staticMembers}
}
            `

            return `
open external class EventEmitter${karakum.ifPresent(heritageClauses, it => ` : ${it}`)} {
${members}\n${companionObject}
}
            `
        }

        return null
    },

    generate(context, render) {
        return []
    }
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/EventEmitterPlugin.kt
package node.karakum.plugins

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Plugin
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationsConfiguration
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.convertParameterDeclarations
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.impure
import arrow.core.raise.nullable
import typescript.InterfaceDeclaration
import typescript.Node
import typescript.SyntaxKind
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isComputedPropertyName
import typescript.isConstructorDeclaration
import typescript.isGetAccessorDeclaration
import typescript.isIdentifier
import typescript.isIndexSignatureDeclaration
import typescript.isInterfaceDeclaration
import typescript.isIntersectionTypeNode
import typescript.isMethodDeclaration
import typescript.isMethodSignature
import typescript.isModuleBlock
import typescript.isModuleDeclaration
import typescript.isParameter
import typescript.isPropertyDeclaration
import typescript.isQualifiedName
import typescript.isSetAccessorDeclaration
import typescript.isTypeReferenceNode

private fun extractModifiers(member: Node) = when {
    isPropertyDeclaration(member) -> member.modifiers
    isMethodDeclaration(member) -> member.modifiers
    isConstructorDeclaration(member) -> member.modifiers
    isGetAccessorDeclaration(member) -> member.modifiers
    isSetAccessorDeclaration(member) -> member.modifiers
    isIndexSignatureDeclaration(member) -> member.modifiers
    else -> null
}?.asArray() ?: emptyArray()

private fun isEventEmitterInterfaceNode(node: Node) = nullable {
    ensure(isInterfaceDeclaration(node))
    ensure(node.name.text == "EventEmitter")

    val moduleBlock = ensureNotNull(node.getParentOrNull())
    ensure(isModuleBlock(moduleBlock))

    val moduleDeclaration = ensureNotNull(moduleBlock.getParentOrNull())
    ensure(isModuleDeclaration(moduleDeclaration))

    val moduleName = moduleDeclaration.name
    ensure(isIdentifier(moduleName))
    ensure(moduleName.text == "NodeJS")
} != null

private fun  isEventEmitterClassNode(node: Node) = nullable {
    ensure(isClassDeclaration(node))
    ensure(node.name?.text == "EventEmitter")
} != null

class EventEmitterPlugin : Plugin {
    private lateinit var eventEmitterInterfaceNode: InterfaceDeclaration

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        ensure(isInterfaceDeclaration(node))
        ensure(isEventEmitterInterfaceNode(node))

        eventEmitterInterfaceNode = node
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("events.d.ts"))

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

        nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Key")

            "EventType"
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull(1))
            ensure(isTypeReferenceNode(secondType))

            val typeName = secondType.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "Key2")

            "EventType"
        } ?: nullable {
            ensure(isTypeReferenceNode(node))

            val typeName = node.typeName
            ensure(isIdentifier(typeName))
            ensure(
                typeName.text == "Listener1"
                        || typeName.text == "Listener2"
            )

            "EventListener"
        } ?: nullable {
            ensure(isIdentifier(node))
            ensure(node.text == "eventName")

            val parameter = ensureNotNull(node.getParentOrNull())
            ensure(isParameter(parameter))
            ensure(parameter.name === node)

            val method = ensureNotNull(parameter.getParentOrNull())
            ensure(
                isMethodDeclaration(method)
                        || isMethodSignature(method)
            )

            "type"
        } ?: nullable {
            ensure(isMethodDeclaration(node))

            val classNode = ensureNotNull(node.getParentOrNull())
            ensure(isEventEmitterClassNode(classNode))

            nullable {
                ensure(isComputedPropertyName(node.name))

                "/* ${typeScriptService.printNode(node)} */"
            } ?: nullable {
                val name = next(node.name)

                val returnType = node.type?.let { next(it) }

                convertParameterDeclarations(
                    node, context, next,
                    ParameterDeclarationsConfiguration(
                        strategy = ParameterDeclarationStrategy.function,
                        template = { parameters, signature ->
                            nullable {
                                ensure(signature.any { it.type?.kind == SyntaxKind.SymbolKeyword })

                                ""
                            } ?: nullable {
                                ensure(signature.size == 3)

                                val firstParameter = signature.first()

                                val firstParameterType = ensureNotNull(firstParameter.type)
                                ensure(isTypeReferenceNode(firstParameterType))

                                val firstParameterTypeName = firstParameterType.typeName
                                ensure(isQualifiedName(firstParameterTypeName))
                                ensure(firstParameterTypeName.right.text == "EventEmitter")

                                val enhancedReturnType = if (name == "once") "Promise<P>" else "AsyncIterator<P>"

                                "fun <P : Tuple> ${name}(emitter: EventEmitter, type: EventType, options: StaticEventEmitterOptions = definedExternally): $enhancedReturnType"
                            } ?: nullable {
                                ensure(signature.size == 3)

                                val firstParameter = signature.first()

                                val firstParameterType = ensureNotNull(firstParameter.type)
                                ensure(isTypeReferenceNode(firstParameterType))

                                val firstParameterTypeName = firstParameterType.typeName
                                ensure(isIdentifier(firstParameterTypeName))
                                ensure(firstParameterTypeName.text == "EventTarget")

                                val enhancedReturnType = if (name == "once") "Promise<Tuple1<E>>" else "AsyncIterator<Tuple1<E>>"

                                "fun <E : Event> ${name}(emitter: EventTarget, type: web.events.EventType<E>, options: StaticEventEmitterOptions = definedExternally): $enhancedReturnType"
                            } ?: nullable {
                                ensure(signature.size == 2)

                                val firstParameter = signature.first()

                                val firstParameterType = ensureNotNull(firstParameter.type)
                                ensure(isTypeReferenceNode(firstParameterType))

                                val firstParameterTypeName = firstParameterType.typeName
                                ensure(isQualifiedName(firstParameterTypeName))
                                ensure(firstParameterTypeName.right.text == "EventEmitter")

                                "fun ${name}(emitter: EventEmitter, type: EventType)${ifPresent(returnType, { ": $it" })}"
                            } ?: nullable {
                                ensure(signature.size == 2)

                                val firstParameter = signature.first()

                                val firstParameterType = ensureNotNull(firstParameter.type)
                                ensure(isTypeReferenceNode(firstParameterType))

                                val firstParameterTypeName = firstParameterType.typeName
                                ensure(isIdentifier(firstParameterTypeName))
                                ensure(firstParameterTypeName.text == "EventTarget")

                                "fun ${name}(emitter: EventTarget, type: web.events.EventType<*>)${ifPresent(returnType, { ": $it" })}"
                            } ?: run {
                                // remove generics
                                "fun ${name}(${parameters})${ifPresent(returnType, { ": $it" })}"
                            }
                        }
                    )
                )
            }
        } ?: nullable {
            ensure(isMethodSignature(node))

            val interfaceNode = ensureNotNull(node.getParentOrNull())
            ensure(isEventEmitterInterfaceNode(interfaceNode))

            nullable {
                ensure(isComputedPropertyName(node.name))

                "/* ${typeScriptService.printNode(node)} */"
            } ?: nullable {
                val name = next(node.name)

                val returnType = node.type?.let { next(it) }

                convertParameterDeclarations(
                    node, context, next,
                    ParameterDeclarationsConfiguration(
                        strategy = ParameterDeclarationStrategy.function,
                        template = { parameters, _ ->
                            // remove generics
                            "fun ${name}(${parameters})${ifPresent(returnType, { ": $it" })}"
                        }
                    )
                )
            }
        } ?: nullable {
            ensure(isClassDeclaration(node))
            ensure(node.name?.text == "EventEmitter")

            val heritageClauses = node.heritageClauses?.asArray()
                ?.joinToString(", ") { next(it) }

            val members = node.members.asArray()
                .filter { member -> extractModifiers(member).all { it.kind != SyntaxKind.StaticKeyword } }
                .plus(eventEmitterInterfaceNode.members.asArray())
                .joinToString("\n") { next(it) }

            val staticMembers = node.members.asArray()
                .filter { member -> extractModifiers(member).any { it.kind == SyntaxKind.StaticKeyword } }
                .joinToString("\n") { next(it) }

            val companionObject = """
                companion object {
                $staticMembers
                }
            """.trimIndent()

            """
                open external class EventEmitter${ifPresent(heritageClauses, { " : $it" })} {
                $members
                $companionObject
                }
            """.trimIndent()
        }
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
